src/subprocess-posix.cc

```diff
@@ -29,6 +29,19 @@ extern char** environ;
 
 #include "util.h"
 
+int HandleEAGAINPosixSpawn(pid_t* pid, const char* path,
+                           const posix_spawn_file_actions_t* file_actions,
+                           const posix_spawnattr_t* attrp, char* const argv[],
+                           char* const envp[]) {
+  while (true) {
+    int ret = posix_spawn(pid, path, file_actions, attrp, argv, envp);
+    if (ret == -1 && errno == EAGAIN)
```

Comment:
Something to be careful of:

posix_spawn returns the same error codes as fork and vfork (Since it calls one of those under the hood).

From: http://man7.org/linux/man-pages/man2/fork.2.html

> 
>        EAGAIN A system-imposed limit on the number of threads was
>               encountered.  There are a number of limits that may trigger
>               this error:
> 
>               *  the RLIMIT_NPROC soft resource limit (set via
>                  setrlimit(2)), which limits the number of processes and
>                  threads for a real user ID, was reached;
> 
>               *  the kernel's system-wide limit on the number of processes
>                  and threads, /proc/sys/kernel/threads-max, was reached (see
>                  proc(5));
> 
>               *  the maximum number of PIDs, /proc/sys/kernel/pid_max, was
>                  reached (see proc(5)); or
> 
>               *  the PID limit (pids.max) imposed by the cgroup "process
>                  number" (PIDs) controller was reached.
> 

So EAGAIN doesn't happen if you create processes too *quickly*. It happens when the number of processes exceeds some limit.

I don't know that simply looping will actually address the problem you want to address.

A more comprehensive solution to this issue is

1) Have a group of tasks that are supposed to be "alive" according to the dependency graph. These tasks can be in the "waiting to launch" state, the "launched" state, and the "finished" state.
2) When a new task is added to that group, attempt to launch it immediately, if possible.
3) When a task enters the finished state, attempt to launch other tasks that are in the "waiting to launch" state, in addition to the other current activities that Ninja would do (e.g. add the next set of tasks to the list of child processes)
4) Periodically (some user and/or system specified time period) attempt to launch one or more of the "waiting to launch" tasks.
5) Tasks attempting to transition from "waiting to launch" to "launched" only stay in the "waiting to launch" state if the error code from posix_spawn indicates that attempting to launch the task again later can possibly succeed.

Reply:
Yeah, I had started to think about whether it would make any sense to methodically lower the number of concurrent processes allowed whenever EAGAIN is specified, altering the maximum parallelism ninja feels like it can utilize. This would require the command runner code to have a more flexible return code that would effectively re-queue a task to be run whenever a process "resource" is available. That breaks some nice encapsulation in the Edge/Node code. And, it is indeed more complex. At some point the rubber would have to meet the road and we'd have to decide whether ninja should be allowed to keep trying to fork or not. Going to have to think on it a bit. I'm open to more ideas, keep 'em coming... Again, thanks for the feedback.